### ISSUE TITLE
[Question] Would a schema file make sense?

### ISSUE THREAD

**Issue by eimermusic** [laravel/internals#980]
Hi,
Coming back to PHP from Rails land after many years I was curious about one "missing" aspect of the Laravel migration feature.

This is schema file saved after each migration that allows a new environment to short-circuit the many migrations by simply importing the "current" state (and fills the migrations table with the rows declaring that all known migrations have been run).

The main benefits is speed and being less error prone. 

Personally, the protection against errors is my primary concern. When an app ages you can end up having some code in an ancient migration that fails because the methods it uses to do some data manipulation might have been deprecated or refactored long ago. It is not fun realizing that a method or function you deleted 3 years ago is preventing a clean setup from migrating forward to the current state.

Any particular reason, other than priority and effort, that a schema file feature is not in Laravel?

I would be interested in working on this, but not if there is some reason against it that I am not seeing.

thanks in advance for any comments

**Comment by mfn** [comment]
_Somehow_ related: https://github.com/laravel/internals/issues/944

This other issue talks about "diffing" and "per-table" schema which is already a step away from pure delta-files and more to what you have in mind with a single schema for all changes.

Especially your suggestion:
> This is schema file saved after each migration

@sisve [mentioned](https://github.com/laravel/internals/issues/944#issuecomment-356296963) that https://github.com/laravel-doctrine/orm can do some of the things, notably from https://laraveldoctrine.org/docs/1.3/migrations/diff 
> The command generates a migration by comparing project current database to mapping information. Doctrine provides this command to generate migration classes by changing entity mappings instead of manually adding modifications to migration class.

No quite what your goal is, though.

**Comment by sisve** [comment]
What's the use-case here? A developer that has a schema file (previously committed to git?) and a blank database, and needs to get up-to-date quickly?

I see some problems with using this in production since migrations can do so much more than just schema changes, they are often used to update/migrate data to new formats too. A schema file wouldn't be able to handle this. It would only work in development when you're executing a seeder to do an initial population afterwards.

**Comment by drbyte** [comment]
> Personally, the protection against errors is my primary concern. When an app ages you can end up having some code in an ancient migration that fails because the methods it uses to do some data manipulation might have been deprecated or refactored long ago. It is not fun realizing that a method or function you deleted 3 years ago is preventing a clean setup from migrating forward to the current state.

I get your point. And I wonder if the solution is as simple as a mindset shift?  For example, I'm in the habit of ensuring that any refactoring done to my Laravel apps includes updating old migration code ... so in the case of deleting an old function, that's already addressed as part of the comprehensive attention to refactoring.

**Comment by tomschlick** [comment]
I have done this in previous projects that had hundreds of migrations and its definitely needed. Instead of saving it on every migration, we had a `migrate:cache` call that would regenerate the db and cache the schema + the migrations table contents. 

That way you can easily run any new migrations in your day to day deployment but developer machines or new production environments don't have to run all the migrations from the beginning of time. Super useful for multi-db multi-tenant setups.

**Comment by drbyte** [comment]
> Super useful for multi-db multi-tenant setups.

Ya, I can see the benefit of that at scale.


> we had a migrate:cache call that would regenerate the db and cache the schema + the migrations table contents.

Any chance you can post that code, even if just as a gist, if not turn it into a package?

**Comment by tomschlick** [comment]
@drbyte yeah I was going to include it in https://github.com/tomschlick/townhouse (multi-tenant db package im working on) but it might make sense to keep it its own package and just include it there.

**Comment by eimermusic** [comment]
@tomschlick It does sound like the migrate:cache command you had is pretty much what I am talking about. I would also be interested in looking at that, even in a rough form.

**Comment by ahinkle** [comment]
I agree here. My migrations file can turn out to be a mess when adding a column, removing, etc. it would be nice to keep it to one schema file that we can modify and change easily.

**Comment by mbitokhov** [comment]
I was thinking about the use cases of this. And while I don't think we should completely get rid of migrations because of the issues brought up by @sisve. Laravel should at least provide something like it to make query building easier.

Laravel already has something similar (eg: $fillable and $date in model). We could extend the concepts but put them in a different file. And one of the major benefits would be in query building to use it to figure out relationships dynamically instead of having it be predefined by the developer

I'll try to experiment with the concept in my free time

**Comment by taylorotwell** [comment]
If someone has a platform agnostic schema file generator I would definitely be interested in looking into including something like this in the core.

**Comment by eimermusic** [comment]
I don't have a schema generator. Given how much I stumble in Laravel still, I feel ill equipped to try writing one too. Sorry.

Referencing the structure.sql file in a Rails project I have available to me, they seem to lean on the command-line tools for the database engine that is configured. This does seem to assume you are using the same DB engine in dev and production... or that you only leverage the schema file in the dev environment.

E.g: "mysqldump --no-data" + appending the table contents dump for just the migrations table.


Reading back a few posts I wanted to clarify some of the use-case details that might not be clear. In case this can be for someones benefit:
• The schema would be generated after every migration and include all current rows of the migrations table (so that, after import, the table contains what has been migrated already)
• You only ever import this schema explicitly by a "schema load" command.
• Anyone could choose to bootstrap their setup using migrate rather than a schema load if they preferred.
• You only ever do this to an empty database. There is no consideration for data modifications at this point.
• Also, a populated db would complain that tables exist already. The schema load command should probably check before importing rather than relying on errors being thrown.
• Since a production database is never (intentionally) wiped and started over the feature would in most cases be used for development, tests, CI, and such.

**Comment by staudenmeir** [comment]
I've started looking into implementing this feature and was thinking about the different file formats.

By default, [Rails](https://edgeguides.rubyonrails.org/active_record_migrations.html#what-are-schema-files-for-questionmark) creates a Ruby file ("one very big migration"). You can also change the format and create an SQL dump file.

AFAICS, we could just take the same approach in Laravel. But I was wondering about the advantages/necessity of a PHP file, i.e. a list of `Schema::create();` statements.

Some advantages of a raw SQL file I see:
- It's *much* easier to implement the file generator, you can just dump the database structure.
- The import into the database is faster.
- It supports all the (obscure) features that users can create with Laravel migrations (e.g. generated columns) or with raw statements (e.g. triggers).

Why we (might) also need a PHP file:
- When the user wants to use the schema file to get a quick overview of their database structure, a PHP file would be much more readable than an SQL file.
**But:** Can't the user achieve the same goal with their favorite administration tool (e.g. phpMyAdmin)?
- The user can populate different databases from a single schema file.
**But:** This smells like people who are using SQLite to test their MySQL applications. Besides that, you can't always reliably reverse-engineer a migration file from a database: `$table->uuid()` creates a `char` column on MySQL. So if you generate a schema file from this database and run it on PostgreSQL (which has a native `uuid` column type), you don't get the expected result.
- The user lacks the permissions to create a dump file (e.g. can't access `pg_dump` on PostgreSQL).
**But:** Since the whole feature is only intended for development machines, this shouldn't be a problem.

**Comment by mfn** [comment]
It seems you yourself favor the SQL file.

I'd say it's sufficient for the start.

I'm wondering though: how would you generate the SQL itself?

**Comment by staudenmeir** [comment]
I started by implementing the PHP file generator and it went quite well, as the `doctrine/dbal` package does most of the work. But it doesn't support all the Laravel migration features (natively).

So the more limitations I found, the more I thought about the advantages of a raw SQL file.

On MySQL, you can use `SHOW CREATE TABLE`. The other databases have similar functionalities.

**Comment by michaeldyrynda** [comment]
If you have an SQL file, you’re tied to the database engine that generated it. Whilst it’s not likely, with migrations, you could conceivably move between SQLite, MySQL, and Postgres over time. 

With an SQL file, you’d have no way of getting between each necessarily. 

In our projects with legacy databases where we _know_ we’re always going to use MySQL, we generated an SQL dump and put that through `DB::unprepared`, then all additions from there are theough discrete migrations. 

Allows you to neatly separate legacy from new changes, too.

**Comment by staudenmeir** [comment]
@michaeldyrynda As I mentioned, it's not always possible to reverse-engineer a migration file that works on different databases.

**Comment by mfn** [comment]
> I started by implementing the PHP file generator and it went quite well, as the `doctrine/dbal` package does most of the work.

Interesting. And how "compatible" is the result?

Talking about (mostly Postgres, that's what I'm using) thing like:
- partial indices
- triggers / procedures
- even more special cases like `ALTER TABLE` with specific syntax, like `SET (FILLFACTOR=…)`

I was just quickly skimming in a project through the `unprepared` calls in a migration of mid-sized project here.

**Comment by michaeldyrynda** [comment]
Not being snake to generate cross-provider SQL is kind of a big negative imo. I don’t think the framework should offer anything that provides less flexibility than it has now.

**Comment by staudenmeir** [comment]
@michaeldyrynda I think there's a misunderstanding here. The schema file we are discussing would only be an addition to the regular migrations. People could use it to quickly populate their development/testing databases without running all the individual migrations.

**Comment by michaeldyrynda** [comment]
Eh, honestly, I would just generate the SQL file as needed by hand and throw it into `DB::unprepared()`

**Comment by sisve** [comment]
> Besides that, you can't always reliably reverse-engineer a migration file from a database: $table->uuid() creates a char column on MySQL. So if you generate a schema file from this database and run it on PostgreSQL (which has a native uuid column type), you don't get the expected result.

When I run `$table->uuid()` I expect a column that could store an uuid. Preferably a column specialized for the case, but a varchar column works too. The important part is that I can run `WHERE column = 'my-guid-zomg'` and similar statements. It's my responsibility to pick a database that can handle my use-cases, including knowing if the uuid storage is optimized enough for my case.

We could use column comments to keep type information. That uuid column could be a varchar with the comment `(DC2Type:uuid)`, and from that comment we know that this was meant as an uuid column. We could also support other things, like `(DC2Type:json)` for json columns on databases without native json support.

Those comments allow any new custom types to be implemented. We could have a `(DC2Type:carbon)` that knows how to build datetime columns, and in the long run be used to hydrate these back into Carbon objects instead of DateTime objects. This would fit into the recurring discussion about custom casts; an extendible type system where we could add more types that handles the conversion between database values and php values. How about a [(DC2Type:rot13)](https://github.com/doctrine/doctrine2/blob/master/tests/Doctrine/Tests/DbalTypes/Rot13Type.php)?

Doctrine has already solved these things.

* Doctrine has schema files (xml, yaml and probably more)
* Doctrine can validate the schemas for internal consistency (relations between classes)
* Doctrine uses column comments to remember meta-information
* Doctrine can read the database and verify it against the current schema file
* Doctrine can use the schema file to create a new database, or modify the existing one to match

I feel that an attempt to make this ourselves will result in a huge chunk of code that would be hard to maintain, and an eternal stream of bug reports of special cases. Why not let another project handle that?
